apps/gitness/src/components/FileEditor.tsx

```diff
@@ -0,0 +1,271 @@
+  ToggleGroup,
+  ToggleGroupItem
+} from '@harnessio/canary'
+import { SandboxLayout } from '@harnessio/playground'
+import { useGetContentQuery } from '@harnessio/code-service-client'
+import { useGetRepoRef } from '../framework/hooks/useGetRepoPath'
+import {
+  decodeGitContent,
+  filenameToLanguage,
+  normalizeGitRef,
+  GitCommitAction,
+  FILE_SEPERATOR,
+  PLAIN_TEXT
+} from '../utils/git-utils'
+import { PathParams } from '../RouteDefinitions'
+import { PathParts, splitPathWithParents } from '../utils/path-utils'
+import { CodeDiffEditor, CodeEditor } from '@harnessio/yaml-editor'
+import { themes } from '../pages/pipeline-edit/theme/monaco-theme'
+import GitCommitDialog from './GitCommitDialog'
+
+export type ViewTypeValue = 'contents' | 'changes'
+
+export const FileEditor: React.FC = () => {
+  const repoRef = useGetRepoRef()
+  const { spaceId, repoId, gitRef, resourcePath } = useParams<PathParams>()
+  const subResourcePath = useParams()['*'] || ''
+  const repoPath = `/${spaceId}/repos/${repoId}/code/${gitRef}`
+  const fullResourcePath = subResourcePath ? resourcePath + '/' + subResourcePath : resourcePath
+  const [fileName, setFileName] = useState<string>()
+  const [language, setLanguage] = useState('')
+  const [originalFileContent, setOriginalFileContent] = useState('')
+  const [content, setContent] = useState(originalFileContent)
+  const [view, setView] = useState<ViewTypeValue>('contents')
+  const [dirty, setDirty] = useState(false)
+  const [isCommitDialogOpen, setIsCommitDialogOpen] = useState(false)
+  const navigate = useNavigate()
+
+  const { data: repoDetails } = useGetContentQuery({
+    path: fullResourcePath || '',
+    repo_ref: repoRef,
+    queryParams: { include_commit: true, git_ref: normalizeGitRef(gitRef || '') }
+  })
+
+  const themeConfig = useMemo(
+    () => ({
+      defaultTheme: 'dark',
+      themes
+    }),
+    []
+  )
+
+  const isNew = useMemo(() => repoDetails && (!repoDetails?.content?.data || repoDetails.type === 'dir'), [repoDetails])
+  const [parentPath, setParentPath] = useState(
+    isNew ? fullResourcePath : fullResourcePath?.split(FILE_SEPERATOR).slice(0, -1).join(FILE_SEPERATOR)
+  )
+  const pathParts = useMemo(() => (parentPath?.length ? splitPathWithParents(parentPath) : []), [parentPath])
+  const fileResourcePath = useMemo(
+    () => [(parentPath || '').trim(), (fileName || '').trim()].filter(p => !!p.trim()).join(FILE_SEPERATOR),
+    [parentPath, fileName]
+  )
+  const isUpdate = useMemo(() => fullResourcePath === fileResourcePath, [fullResourcePath, fileResourcePath])
+  const commitAction = useMemo(
+    () => (isNew ? GitCommitAction.CREATE : isUpdate ? GitCommitAction.UPDATE : GitCommitAction.MOVE),
+    [isNew, isUpdate]
+  )
+  const inputRef = useRef<HTMLInputElement | null>()
+
+  useEffect(() => {
+    if (inputRef.current && fileName?.length) {
+      inputRef.current.size = Math.min(Math.max(fileName?.length - 2, 20), 50)
+    }
+  }, [fileName, inputRef])
```

Comment:
if u are editing the filename, does this cause a rerender?

Reply:
yes it does